File: 백준/Gold/16235. 나무 재테크/나무 재테크.cc
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int dir[8][2] = {{-1, 0}, {-1, 1}, {0, 1}, {1, 1}, {1, 0}, {1, -1}, {0, -1}, {-1, -1}};
int N, M, K;

vector<vector<int>> board;
vector<vector<int>> add_board;
vector<vector<vector<int>>> trees;

void spring()
{
    for (int i = 0; i < N; ++i)
    {
        for (int j = 0; j < N; ++j)
        {
            if (!trees[i][j].empty())
            {
                sort(trees[i][j].begin(), trees[i][j].end());

                vector<int> del_list;
                for (int t = 0; t < trees[i][j].size(); ++t)
                {
                    if (board[i][j] >= trees[i][j][t])
                    {
                        board[i][j] -= trees[i][j][t];
                        trees[i][j][t] += 1;
                    }
                    else
                    {
                        del_list.push_back(t);
                    }
                }
                for (int d = del_list.size() - 1; d >= 0; --d)
                {
                    board[i][j] += trees[i][j][del_list[d]] / 2;
                    trees[i][j].erase(trees[i][j].begin() + del_list[d]);
                }
            }
        }
    }
}

void fall()
{
    for (int i = 0; i < N; ++i)
    {
        for (int j = 0; j < N; ++j)
        {
            if (!trees[i][j].empty())
            {
                for (int t : trees[i][j])
                {
                    if (t % 5 == 0)
                    {
                        for (int d = 0; d < 8; ++d)
                        {
                            int nx = i + dir[d][0];
                            int ny = j + dir[d][1];

                            if (0 <= nx && nx < N && 0 <= ny && ny < N)
                            {
                                trees[nx][ny].push_back(1);
                            }
                        }
                    }
                }
            }
        }
    }
}

void winter()
{
    for (int i = 0; i < N; ++i)
    {
        for (int j = 0; j < N; ++j)
        {
            board[i][j] += add_board[i][j];
        }
    }
}

int main()
{
    cin >> N >> M >> K;

    board.assign(N, vector<int>(N, 5));
    add_board.assign(N, vector<int>(N));
    trees.assign(N, vector<vector<int>>(N));

    for (int i = 0; i < N; ++i)
    {
        for (int j = 0; j < N; ++j)
        {
            cin >> add_board[i][j];
        }
    }

    for (int i = 0; i < M; ++i)
    {
        int x, y, year;
        cin >> x >> y >> year;
        trees[x - 1][y - 1].push_back(year);
    }

    for (int i = 0; i < K; ++i)
    {
        spring();
        fall();
        winter();
    }

    int result = 0;
    for (int i = 0; i < N; ++i)
    {
        for (int j = 0; j < N; ++j)
        {
            result += trees[i][j].size();
        }
    }

    cout << result << endl;

    return 0;
}

File: 백준/Gold/16235. 나무 재테크/나무 재테크.py
import copy

dir = [(-1, 0), (-1, 1), (0, 1), (1, 1), (1, 0), (1, -1), (0, -1), (-1, -1)]
N, M, K = map(int, input().split())

board = [[5] * N for _ in range(N)]
add_board = [list(map(int, input().split())) for _ in range(N)]
trees = [[list() for _ in range(N)] for _ in range(N)]

for _ in range(M):
    x, y, year = map(int, input().split())
    trees[x - 1][y - 1].append(year)


def spring():
    for i in range(N):
        for j in range(N):
            if trees[i][j]:
                trees[i][j].sort()

                del_list = []
                for t in range(len(trees[i][j])):
                    if board[i][j] >= trees[i][j][t]:
                        board[i][j] -= trees[i][j][t]
                        trees[i][j][t] += 1
                    else:
                        del_list.append(t)
                for d in del_list[::-1]:
                    board[i][j] += trees[i][j][d] // 2
                    del trees[i][j][d]


def fall():
    for i in range(N):
        for j in range(N):
            if trees[i][j]:
                for t in trees[i][j]:
                    if t % 5 == 0:
                        for dx, dy in dir:
                            nx, ny = i + dx, j + dy

                            if 0 <= nx < N and 0 <= ny < N:
                                trees[nx][ny].append(1)


def winter():
    for i in range(N):
        for j in range(N):
            board[i][j] += add_board[i][j]


for _ in range(K):
    spring()
    fall()
    winter()

result = 0
for i in range(N):
    for j in range(N):
        for t in trees[i][j]:
            if t != 0:
                result += 1

print(result)
